capi-configuration/README.md

```diff
@@ -0,0 +1,333 @@
+  name: {{clusterID}}-mp-0
+spec:
+  clusterName: {{clusterID}}
+  template:
+    spec:
+      bootstrap:
+        configRef:
+          name: {{clusterID}}-mp-0
+      infrastructureRef:
+        name: {{clusterID}}-mp-0
+```
+This template is going to be stable for any `v1alpha3` cluster and would only need minimal changes to be stable for `v1alpha4`.
+
+### Cluster-api base files
+
+We need to default a big set of custom resources with `cluster-api`.
+These custom resources need to be created and defaulted in a consistent way.
+The approach we follow here is to have a single source of truth which is as complete as possible.
+
+`base files` are our static defaulting values - any webhook or tool which does defaulting should use them as the source of truth.
+
+Let's look at an example to clarify the technical implications.
+A valid `MachinePool` CR in `v1alpha3` can look like this:
+```yaml
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: MachinePool
+metadata:
+  name: capa-mp-0
+spec:
+  clusterName: capa
+  replicas: 2
+  template:
+    spec:
+      bootstrap:
+        configRef:
+          apiVersion: bootstrap.cluster.x-k8s.io/v1alpha3
+          kind: KubeadmConfig
+          name: capa-mp-0
+      clusterName: capa
+      infrastructureRef:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+        kind: AWSMachinePool
+        name: capa-mp-0
+      version: v1.16.8
+```
+From this custom resource we can now extract the values which are static and configurable.
+So values which a customer might want to change the default of _and_ which we can default statically independent of any other content in the custom resource.
+
+In this particular case there are two values which are both static and configurable:
+- `spec.replicas`
+- `spec.template.spec.version`
+
+We can now construct a `ConfigMap` with these values which are then going to be used by any defaulting tool/webhook:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: machinepool-21-0-0
+  namespace: giantswarm
+data:
+  spec:
+    replicas: 3
+    template:
+      spec:
+        version: v1.21.4
+```
+It is important to note that a assumption is that **no** values in the base file are immutable!
+
+### Cluster-api context dependent values
+
+We have some cases in `cluster-api` where a value in a custom resource is neither part of the fundamental structure nor is it easy to statically default.
+
+These values are usually highly dependent on the context in which the cluster is created in.
+For example the `CIDR` in which the cluster is created.
+We currently assume that these values will be defaulted by custom webhooks or operators with significant provider specific domain knowledge.
+
+### Putting together a Giant Swarm release
+
+A Giant Swarm releases is just a set of `cluster-api` base files as defined previously with added `apps`.
+
+The following folder structure would be contained in `giantswarm/example-customer`:
+```yaml
+aws:
+  v21.0.0:
+    base:
```

Comment:
Why are `base` defaults not defined globally (like in `releases`)? Why per-customer?

Reply:
In my mind there is a global definition in a GS repository which gets replicated to customer repositories. This is because templating tools such as `kustomize` can not reason across repositories AFAIK

Reply:
I think this is a global one, just that we would distribute it (still think that's 2nd phase) to customer specific repos, to then be able to use the customer specific repo as an interface of transparency to the customer, this way they also would not need to check back with the source of truth just to see the base settings.

Reply:
Ok I see. If it is a technical limitation we can work around, I'd personally be inclined to not duplicate the source of truth, but would be worth getting some other opinions on if it's that much better to have them in the customer repo. Having only theoretically immutable resources in a bunch of shared repos just seems like begging for some well-meaning person to modify them